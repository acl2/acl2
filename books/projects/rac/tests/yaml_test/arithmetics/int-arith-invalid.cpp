#include <ac_int.h>
#include <array>

using namespace std;

// RAC begin

int binary_invalid(int a, array<int, 3> b) { return a && b; }

int prefix_invalid() {
  array<int, 3> a = { 0, 1, 2 };
  return -a;
}

int no_implicit_cast() {
  ac_int<3, false> x = 0;
  array<int, 3> a = { 0, 1, 2 };
  x = x + a;

  return 0;
}

int ternary_invalid() {
  ac_int<3, false> x = 0;
  array<int, 3> a = { 0, 1, 2 };
  x = true ? x : a;

  return 0;
}

typedef array<int, 3> array3;
int prefix_invalid_with_typedef() {
  array3 a = { 0, 1, 2 };
  return -a;
}

// RAC end
