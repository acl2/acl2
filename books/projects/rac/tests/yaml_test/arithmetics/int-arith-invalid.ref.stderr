int-arith-invalid.cpp:8 (52-58): Cannot apply `&&` to arguments of type int and int[3]
  8 | int binary_invalid(int a, array<int, 3> b) { return a && b; }
    |                                                     ^^^^^^
note: They should be both a primitive type (like an int or enum) or a register type.

int-arith-invalid.cpp:12 (10-11): Cannot apply `-` to an argument of type `int[3]` which is not a register type or a primitive type, aka int, int64, uint, uint64, bool.
 12 |   return -a;
    |           ^

int-arith-invalid.cpp:18 (10-11): Invalid type: right operand is a `ac_int<3, false>` but left (int[3]) is not convertible to a register type.
 18 |   x = x + a;
    |           ^

int-arith-invalid.cpp:26 (6-18): Left and right do not share same type (left is `ac_int<3, false>` and right `int[3]`).
 26 |   x = true ? x : a;
    |       ^^^^^^^^^^^^

int-arith-invalid.cpp:34 (10-11): Cannot apply `-` to an argument of type `int[3]` which is not a register type or a primitive type, aka int, int64, uint, uint64, bool.
 34 |   return -a;
    |           ^
